feat: new API endpoint /hierarchy.php

foodev opened the pull request:
Possible parameters (and values):

- format: json
- country_code: any valid ISO 3166-1 alpha-2 country code

You will get a list of all regions (boundary=administrative) within the
given country. The list starts with the country itself.

Each record has the key `children` which consists of another lists with
regions etc.

Example usage:

```
https://nominatim.example.org/hierarchy.php?format=json&country_code=de
```

Example response:

```
[
    {
        "osm_type": "R",
        "osm_id": 51477,
        "name": "Deutschland",
        "indexed_date": "2021-09-20T03:30:12+00:00",
        "children": [ ... ]
    }
]
```

---

To test this, the easiest way is to use the [Nominatim Docker repository](https://github.com/mediagis/nominatim-docker/tree/master/3.6) and edit the Dockerfile to use this branch.

If you have executed this command before you might have to remove the `nominatim` image(s) first. Otherwise it will not fetch the latest code from GitHub, because the branch name has not changed (this the Dockerfile has not changed and it will use the cache).

```sh
git clone git@github.com:mediagis/nominatim-docker.git \
&& cd nominatim-docker/3.7 \
&& sed -i 's~curl https://nominatim.org/release/Nominatim-$NOMINATIM_VERSION.tar.bz2 -o nominatim.tar.bz2~git clone --config advice.detachedHead=false --single-branch --branch feat/hierarchy --depth 1 --recursive https://github.com/openPetition/Nominatim Nominatim-$NOMINATIM_VERSION~g; s~tar xf nominatim.tar.bz2~curl https://www.nominatim.org/data/country_grid.sql.gz -o Nominatim-$NOMINATIM_VERSION/data/country_osm_grid.sql.gz~g; s~&& rm /app/\*.tar.bz2 ~~g' Dockerfile \
&& docker build -t nominatim . \
&& cd - \
&& rm -rf nominatim-docker \
&& docker run -it --rm \
    -e PBF_URL=https://download.geofabrik.de/europe/malta-latest.osm.pbf \
    -e REPLICATION_URL=https://download.geofabrik.de/europe/malta-updates/ \
    -e IMPORT_STYLE=extratags \
    -p 8080:8080 \
    --name nominatim \
    nominatim
```

Now you should be able to retrieve the hierarchy of Malta via http://localhost:8080/hierarchy.php?format=json&country_code=mt

Commit:
feat: new API endpoint /hierarchy.php

Possible parameters (and values):

- format: json
- country_code: any valid ISO 3166-1 alpha-2 country code

You will get a list of all regions (boundary=administrative) within the
given country. The list starts with the country itself.

Each record has the key `children` which consists of another lists with
regions etc.

Example usage:

```
https://nominatim.example.org/hierarchy.php?format=json&country_code=de
```

Example response:

```
[
    {
        "osm_type": "R",
        "osm_id": 51477,
        "name": "Deutschland",
        "indexed_date": "2021-09-20T03:30:12+00:00",
        "children": [ ... ]
    }
]
```

Commit:
fix: add new hierarchy.php into website script lists

Commit:
fix: paths

Commit:
imp: all regions should be part of `place_addressline`

See documentation for the config file:
https://nominatim.org/release-docs/latest/develop/Ranking/#rank-configuration

foodev commented on lib-php/Hierarchy.php:
```diff
@@ -0,0 +1,148 @@
+                ':osmId' => $iOsmId,
+                ':adminLevel' => $iAdminLevel,
+            ]);
+        }
```
Ich würde gerne diese beiden Abfragen in eine tun. Weiß aber nicht wie...

jtopenpetition replied:
die beiden Queries sind bis auf die SELECT clause ziemlich unterschiedlich. Diese könnte man in nen extra String packen und dann zusammen`.`fügen, ansonsten sehe ich keinen echten Mehrgewinn.

foodev commented on settings/address-levels.json:
```diff
@@ -110,16 +110,16 @@
       },
       "boundary" : {
-          "administrative5" : [10, 0]
+          "administrative5" : 10
```
Als ich das zuerst getestet hatte, hatte ich den gesamten Block für `de` entfernt. Wenn ich aber die [Dokumentation](https://nominatim.org/release-docs/latest/develop/Ranking/#rank-configuration) richtig verstehe, sollte es reichen, die 0 für "address rank" anzupassen. Das habe ich hier also gemacht.

Ich würde das gerne, zumindest für Deutschland, nochmal genauer testen, nur dauert der Import lokal ca. zwei Tage. Ich könnte es also Freitagabend nochmal starten und dann am Montag testen.

jtopenpetition replied:
hast du das nochmal getestet? in jedem Fall sollte das aber helfen, so wie du schon beschreibst, weiß gerade nicht, wie ich das testen soll ohne einen mordsmäßig großen Import zu machen ^^

foodev replied:
> hast du das nochmal getestet?

Nein, ich hatte es leider vergessen...

foodev commented:
Im Moment läuft auf unserem Server Version 3.6. Ich denke es würde Sinn machen, das mal auf 3.7.x zu aktualisieren, da wir wohl eh die Datenbank neu importieren müssen (siehe Änderungen in `settings/address-levels.json`).

Deswegen basiert dieser Branch auf `3.7.x`. Es fühlt sich zwar komisch an, dass dann in den Branch `3.7.x` zu mergen, aber der `master` hat schon zu viele neuere Commits...

Commit:
refactor: use JOIN instead of comma style

jtopenpetition commented on lib-php/Hierarchy.php:
```diff
@@ -0,0 +1,150 @@
+                    AND p1.class = 'boundary'
+                    AND p1.type = 'administrative'
+                    AND p1.osm_type = 'R'
+                    AND p1.admin_level > CASE
+                        WHEN p1.country_code = 'mt' THEN 3
+                        ELSE 2
+                        END
```
das funktioniert, aber auf lange Sicht ist es womöglich schöner `and p1.admin_level > :adminLevel` zu haben und die Logik hierfür im Code zu haben ... womöglich? dann könnte man einfach einen Großteil der Logik in eine Klassenkonstante auslagern und dann beim `getAll` `':adminLevel' => self::FIRST_ADMIN_LEVEL[$sCountryCode] ?? 3` (oder mit isset...)

jtopenpetition commented on lib-php/Hierarchy.php:
```diff
@@ -0,0 +1,150 @@
+
+        $aReturn = [];
+
+        if ($aResult) {
+            foreach ($aResult as $aResultItem) {
+                $aResultItem['children'] = $this->getChildrenRecursively(
+                    $aResultItem['osm_id'],
+                    $iAdminLevel + 1,
+                    $sCountryCode
+                );
+
+                $aReturn[] = $aResultItem;
+            }
+        } else {
+            $aReturn = $this->getChildrenRecursively(
+                $iOsmId,
+                $iAdminLevel + 1,
+                $sCountryCode
+            );
+        } 
```
ich lese hier heraus, dass die admin-level immer lückenlos sein müssen, halte das für eine gefährliche Annahme...

bin mir nicht 100% sicher, wie das N/A genau zu verstehen ist, aber es ist denkbar, dass mittendrin was fehlt. Wir sollten womöglich spätestens beim Import eine große Änderung der Anzahl an Nodes in der Hierarchie blockieren (force flag oder sowas, um trotzdem zu pushen?)
![image](https://user-images.githubusercontent.com/70581149/135064713-2b6654c7-0b95-486d-be39-7dc6c4883543.png)

jtopenpetition reviewed (changes requested):
find das schon ziemlich cool. kleine Sachen sind noch

jtopenpetition commented on lib-php/Hierarchy.php:
```diff
@@ -0,0 +1,150 @@
+
+        $aReturn = [];
+
+        if ($aResult) {
```
\+ kommentar, damit etwas deutlicher wird, dass hier dann ein level weiter gegangen wird

jtopenpetition reviewed (changes requested):
soweit schon ganz schick! ;o)

ein paar kleine Sachen kannst du noch verbessern, aber insgesamt schon tolle Sache :+1:

Commit:
imp: filter unwanted first-level regions via ISO3166-2

There are a few ugly regions (e.g. R2869330 or R3659532)
which are not actually ordinary administrations of the country
and thus should not be part of the hierarchy.
We try to exclude these using the ISO3166-2 code. The assumption
is that as soon as at least one administration has this code,
all valid administrations of this country also have the code.
Except for such ugly regions. Thus, we only take the regions
with the (correct) code.

In addition, the code was restructered a little and comments were
added.

foodev commented:
An Luxemburg hatte man gesehen, dass es leider doch immer noch Ausnahmen geben könnte. Ich weiß nicht, wie man das am besten alles vorher testen könnte (ein lokaler Import von Europa würde sehr lange dauern).

Vermutlich ist es daher am besten, es so erstmal zu deployen (d.h. Nominatim aus diesem Repository installieren + Datenbank erneut importieren) und dann kann man sich einfacher die verschiedenen Hierarchien im Browser angucken. Sollte man mehr nachbessern müssen, sollte es ja reichen, Nominatim (den Code) zu aktualisieren.

jtopenpetition reviewed (approved):
:+1:

machen wir so, wie du vorgeschlagen hast. erstmal deployen, neuer import, und dann das script anpassen wo es nötig ist.

Commit:
feat: docker-compose file to setup Nominatim for development easily
